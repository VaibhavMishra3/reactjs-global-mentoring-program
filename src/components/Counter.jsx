import React from "react";

class Counter extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: props.initialValue };
  }

  increment = () => {
    this.setState({ count: this.state.count + 1 });
  };

  decrement = () => {
    this.setState({ count: this.state.count - 1 });
  };

  render() {
    return React.createElement(
      "div",
      null,
      React.createElement("p", null, `Count: ${this.state.count}`),
      React.createElement("button", { onClick: this.decrement }, "-"),
      React.createElement("button", { onClick: this.increment }, "+")
    );
  }
}

export default Counter;
