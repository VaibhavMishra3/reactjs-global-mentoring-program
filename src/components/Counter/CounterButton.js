import React from 'react';

const CounterButton = (text, testId, handleClick) => {
  return React.createElement(
    'button',
    {
      'data-testid': testId,
      onClick: () => handleClick()
    },
    text
  );
}

export default CounterButton;
