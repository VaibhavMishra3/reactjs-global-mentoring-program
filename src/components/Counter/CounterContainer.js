import React from 'react';

export const CounterContainer = (children) => {
  return React.createElement(
    'div',
    {},
    ...children
  );
}

export default CounterContainer;
