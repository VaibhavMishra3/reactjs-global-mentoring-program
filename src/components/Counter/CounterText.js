import React from 'react';

const CounterText = (value) => {
  return React.createElement(
    'div',
    {
      'data-testid': 'counter-text',
    },
    value
  );
}

export default CounterText;
