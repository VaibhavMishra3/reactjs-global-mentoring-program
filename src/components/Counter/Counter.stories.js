import React from 'react';
import Counter from './Counter';

export default {
  title: 'Components/Counter',
  component: Counter,
};

export const Default = () => <Counter initialValue={0} />;
