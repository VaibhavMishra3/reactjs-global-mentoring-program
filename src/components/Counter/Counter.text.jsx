import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Counter from "./Counter";

test("renders initial value provided in props", () => {
  const { getByText } = render(<Counter initialValue={10} />);
  expect(getByText("Count: 10")).toBeInTheDocument();
});

test('click event on "decrement" button decrements the displayed value', () => {
  const { getByText } = render(<Counter initialValue={5} />);
  const decrementButton = getByText("-");
  fireEvent.click(decrementButton);
  expect(getByText("Count: 4")).toBeInTheDocument();
});

test('click event on "increment" button increments the displayed value', () => {
  const { getByText } = render(<Counter initialValue={5} />);
  const incrementButton = getByText("+");
  fireEvent.click(incrementButton);
  expect(getByText("Count: 6")).toBeInTheDocument();
});
