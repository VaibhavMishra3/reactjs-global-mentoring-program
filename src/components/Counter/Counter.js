import React from 'react';

import CounterButton from './CounterButton';
import CounterContainer from './CounterContainer';
import CounterText from './CounterText';

class Counter extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      value: +this.props.initialValue || 0
    };
  }

  render() {
    return CounterContainer([
      CounterButton('-', 'counter-decrement', this.decrement),
      CounterText(this.state.value),
      CounterButton('+', 'counter-increment', this.increment)
    ]);
  }

  decrement = () => {
    this.setState(prevState => ({
      ...prevState, value: prevState.value - 1
    }));
  }

  increment = () => {
    this.setState(prevState => ({
      ...prevState, value: prevState.value + 1
    }));
  }

}

export default Counter;
