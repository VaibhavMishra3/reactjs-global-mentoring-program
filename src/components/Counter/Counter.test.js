import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import Counter from './Counter';

describe('Counter', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const counter = renderer
      .create(<Counter initialValue='1' />)
      .toJSON();

    expect(counter).toMatchSnapshot();
  });

  //Test that component renders initial value provided in props
  it('should render initial value', () => {
    render(<Counter initialValue='1' />);

    const text = screen.getByTestId('counter-text')
    expect(text.innerHTML).toBe('1');
  });

  //Test that a click event on "decrement" button decrements the displayed value
  it('should decrement the value', () => {
    render(<Counter initialValue='1' />);

    const decrementButton = screen.getByTestId('counter-decrement');
    fireEvent.click(decrementButton);

    const text = screen.getByTestId('counter-text')
    expect(text.innerHTML).toBe('0');
  });

  //Test that a click event on "increment" button increments the displayed value
  it('should increment the value', () => {
    render(<Counter initialValue='1' />);

    const incrementButton = screen.getByTestId('counter-increment');
    fireEvent.click(incrementButton);

    const text = screen.getByTestId('counter-text')
    expect(text.innerHTML).toBe('2');
  });

});
