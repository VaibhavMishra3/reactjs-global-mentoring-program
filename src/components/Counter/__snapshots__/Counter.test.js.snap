// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match snapshot 1`] = `
<div>
  <button
    data-testid="counter-decrement"
    onClick={[Function]}
  >
    -
  </button>
  <div
    data-testid="counter-text"
  >
    1
  </div>
  <button
    data-testid="counter-increment"
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
